File: app/src/main/java/me/yxp/qfun/MainActivity.java
package me.yxp.qfun;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.util.HashSet;
import java.util.Set;

import io.github.libxposed.service.XposedService;
import me.yxp.qfun.utils.hook.hookstatus.HookStatus;
import me.yxp.qfun.utils.qq.HostInfo;

public class MainActivity extends Activity implements View.OnClickListener {
    private TextView customText;
    private Button btnUsage;
    private Button btnDiscussion;
    private Button btnChangelog;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);

        HookStatus.init(this);
        initView();
        updateActivationStatus();
    }

    private void initView() {
        this.customText = findViewById(R.id.custom_text);
        this.btnUsage = findViewById(R.id.btn_usage);
        this.btnDiscussion = findViewById(R.id.btn_discussion);
        this.btnChangelog = findViewById(R.id.btn_changelog);

        btnUsage.setOnClickListener(this);
        btnDiscussion.setOnClickListener(this);
        btnChangelog.setOnClickListener(this);
    }

    public void gotoTelegram() {
        jump("https://t.me/QFunChannel");
    }

    public void gotoGithub() {
        jump("https://github.com/oneQAQone/QFun");
    }

    public void gotoQQGroup() {
        jump("mqqapi://card/show_pslcard?src_type=internal&version=1&uin=1067198087&card_type=group&source=qrcode");
    }

    public void gotoTelegramChat() {
        jump("https://t.me/QFunChatGroup");
    }

    private void jump(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        startActivity(intent);
    }

    private void updateActivationStatus() {
        boolean isHookEnabled = checkHookStatus();
        customText.setText(isHookEnabled ? "状态：已激活" : "状态：未激活");
    }

    private boolean checkHookStatus() {
        boolean isEnabledByLegacyApi = HookStatus.isModuleEnabled() || HostInfo.isInHostProcess();
        boolean isEnabledByLibXposedApi = checkLibXposedStatus();

        return isEnabledByLegacyApi || isEnabledByLibXposedApi;
    }

    private boolean checkLibXposedStatus() {
        XposedService xposedService = HookStatus.getXposedService().getValue();
        if (xposedService == null) {
            return false;
        }

        Set<String> scope = new HashSet<>(xposedService.getScope());
        scope.retainAll(Set.of(HostInfo.PACKAGE_NAME_QQ, HostInfo.PACKAGE_NAME_TIM));
        return !scope.isEmpty();
    }

    @Override
    public void onClick(View v) {
        // 有点懵了这里，原本想用switch
        // 结果R.id的final被砍了，我还不记得恢复的配置是哪个了
        if (v.getId() == R.id.btn_usage)
            gotoGithub();
        else if (v.getId() == R.id.btn_discussion)
            startDiscussionDialog();
        else if (v.getId() == R.id.btn_changelog)
            Toast.makeText(this, "暂时没有更多哦！", Toast.LENGTH_SHORT).show();
    }


    public void startDiscussionDialog() {
        var builder = new AlertDialog.Builder(this, android.R.style.Theme_DeviceDefault_Light_Dialog_Alert);
        var items = new CharSequence[]{
                "Telegram Channel",
                "Telegram Chat Group",
        };
        builder.setTitle("交流讨论")
        .setItems(items, (dialog, which) -> {
            if (which == 0) gotoTelegram();
            if (which == 1) gotoTelegramChat();
        }).show();
    }
}

File: app/src/main/java/me/yxp/qfun/MainActivity.kt
package me.yxp.qfun

import android.content.ComponentName
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.WindowManager
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.net.toUri
import androidx.core.view.WindowCompat
import me.yxp.qfun.ui.components.dialogs.ConfirmDialog
import me.yxp.qfun.ui.core.theme.QFunTheme
import me.yxp.qfun.ui.pages.home.MainScreen
import me.yxp.qfun.utils.hook.hookstatus.HookStatus
import me.yxp.qfun.utils.qq.HostInfo

@Suppress("DEPRECATION")
class MainActivity : ComponentActivity() {

    private var isActivated by mutableStateOf(false)
    private var frameworkInfo by mutableStateOf("")
    private var isIconVisible by mutableStateOf(true)
    private var showHideConfirmDialog by mutableStateOf(false)

    private val launcherComponent by lazy { ComponentName(this, LauncherActivity::class.java) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        WindowCompat.setDecorFitsSystemWindows(window, false)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)

        HookStatus.init(this)
        updateActivationStatus()
        updateIconState()

        setContent {
            QFunTheme(false) {
                MainScreen(
                    BuildConfig.VERSION_NAME,
                    BuildConfig.VERSION_CODE,
                    isActivated,
                    frameworkInfo,
                    isIconVisible,
                    { handleToggleIcon() },
                    { jump("https://t.me/QFunChannel") },
                    { jump("https://github.com/oneQAQone/QFun") },
                    { jump("mqqapi://card/show_pslcard?src_type=internal&version=1&uin=1067198087&card_type=group&source=qrcode") }
                )

                ConfirmDialog(
                    showHideConfirmDialog,
                    "隐藏图标",
                    "确定要隐藏桌面图标吗？\n\n隐藏后您需要通过模块管理器打开设置。\n\n应用将会重启以刷新状态。",
                    "确定",
                    "取消",
                    { showHideConfirmDialog = false },
                    { showHideConfirmDialog = false; toggleIcon(false) }
                )
            }
        }
    }

    private fun updateIconState() {
        val state = packageManager.getComponentEnabledSetting(launcherComponent)
        isIconVisible = state != PackageManager.COMPONENT_ENABLED_STATE_DISABLED
    }

    private fun handleToggleIcon() {
        if (isIconVisible) showHideConfirmDialog = true else toggleIcon(true)
    }

    private fun toggleIcon(enable: Boolean) {
        val newState =
            if (enable) PackageManager.COMPONENT_ENABLED_STATE_ENABLED else PackageManager.COMPONENT_ENABLED_STATE_DISABLED
        try {
            packageManager.setComponentEnabledSetting(launcherComponent, newState, 0)
            Toast.makeText(this, if (enable) "图标已恢复" else "图标已隐藏", Toast.LENGTH_LONG)
                .show()
            isIconVisible = enable
        } catch (_: Exception) {
        }
    }

    private fun jump(url: String) = startActivity(Intent(Intent.ACTION_VIEW, url.toUri()))

    private fun updateActivationStatus() {
        isActivated = checkHookStatus()
        updateFrameworkInfo()
    }

    private fun checkHookStatus(): Boolean {
        val isEnabledByLegacyApi = HookStatus.isModuleEnabled() || HostInfo.isInHostProcess
        val isEnabledByLibXposedApi = checkLibXposedStatus()
        return isEnabledByLegacyApi || isEnabledByLibXposedApi
    }

    private fun checkLibXposedStatus(): Boolean {
        val xposedService = HookStatus.getXposedService().value ?: return false
        val targetScope = setOf(HostInfo.PACKAGE_NAME_QQ, HostInfo.PACKAGE_NAME_TIM)
        return xposedService.scope.intersect(targetScope).isNotEmpty()
    }

    private fun updateFrameworkInfo() {
        if (!isActivated) {
            frameworkInfo = HookStatus.getHookProviderNameForLegacyApi()
            return
        }
        val xposedService = HookStatus.getXposedService().value
        frameworkInfo = if (xposedService != null) {
            "${xposedService.frameworkName} ${xposedService.frameworkVersion} (${xposedService.frameworkVersionCode}), API ${xposedService.apiVersion}"
        } else HookStatus.getHookProviderNameForLegacyApi()
    }
}
